Doc/regions.md

```diff
@@ -0,0 +1,567 @@
+This object is updated in sync with the underlying variables:
+
+```python
+def frame_example1():
+  x = 1
+  l = sys._getframe()
+  x = 2
+  return l.f_locals
+```
+This behaves differently to `locals` and captures the update to `x`.
+```python
+>>> f = frame_example1()
+>>> f
+{'x': 2, 'l': <frame at 0x7ea265e76f60, file '<stdin>', line 5, code frame_example1>}
+```
+Interesting, it also contains `l` which is the frame object itself.
+This is not the case for `locals()`, which does not capture the dictionary it creates.
+
+Storing the frame object in a variable will cause a reference cycle:
+```python
+>>> f["l"].f_locals
+{'x': 2, 'l': <frame at 0x7ea265e76f60, file '<stdin>', line 5, code frame_example1>}
+>>> f["l"].f_locals["l"].f_locals
+{'x': 2, 'l': <frame at 0x7ea265e76f60, file '<stdin>', line 5, code frame_example1>}
+```
+
+There was an interesting example in [numpy](https://github.com/numpy/numpy/blob/e1bf1d635016a0118f7d77bf8931071a2c74ad20/numpy/matrixlib/defmatrix.py#L1095) that uses the context to affect the evaluation of an expression.
+Here we present the idea as a reverse polish evaluator.
+```python
+>>> x = 1
+>>> y = 2
+>>> calc("4 x y + *")
+```
+The idea is to interpolate the `x` and `y` into the string while evaluating it.
+```
+def calc(str):
+  ctxt = sys._getframe().f_back
+  ops = str.split()
+  stack = []
+  for op in ops:
+    try: 
+      stack.append(int(op))
+    except ValueError:
+      if op == "+":
+        stack.append(stack.pop() + stack.pop())
+      elif op == "*":
+        stack.append(stack.pop() * stack.pop())
+      else:    
+        stack.append(ctxt.f_locals[op])      
+  return stack.pop()
+```
+This is a simple example, but it shows how the frame object can be used to access the local variables in the calling scope.
+```
+>>> x = 1
+>>> y = 2
+>>> calc("4 x y + *")
+12
+```
+
+The numpy library is pretty popular, so this is a good example of how to use the frame object in a way that is not debugging.
+
+
+Note that the frame object is mostly read-only, but does contain an interesting operation to set the executing line number of a frame.
+```C
+static int
+frame_setlineno(PyFrameObject *f, PyObject* p_new_lineno, void *Py_UNUSED(ignored))
+```
+This seems very bad to call from another thread/interpreter.
+
+Ultimately, a frame object should never escape the thread/interpreter that it was created in.
+
+
+
+## Regions in Veronapy.
+
+The previous section illustrates the many temporary objects that Python creates to handle variables.
+To support this, we need our region object model to support complex local structures for indexing into regions.
+To achieve this, we will use three kinds of object ownership:
+
+* Local object 
+* Immutable
+* Regions
+
+The "local objects" are used to represent the many local structures that Python creates such as function objects, frame objects and cells.
+These objects are typically not shared between threads, so normally should not require any region tracking.
```

Comment:
Really? Maybe I am taking this out of context, but function objects are shared between threads all the time. (Agreed frames are not, even if they theoretically could.)

Reply:
Question: are these function objects “morally mutable”? Sure they have a ref count, but are there any other mutations possible that are useful and/or common?

Reply:
I think the most common mutation to function objects is to use function attributes ([PEP 232](https://peps.python.org/pep-0232/) and [docs](https://docs.python.org/3/reference/datamodel.html#user-defined-functions)), and the most common use of those is by decorators, e.g. [`@final`](https://docs.python.org/3/library/typing.html#typing.final) attempts to set an attribute `__final__`. Such decorators are typically applied at function definition time, so once the module is imported, it will typically be fine to freeze the function object.

However, people love to hack, and I'm sure that there are quite a few libraries that mess with writable function attributes. There is no moral prejudice against such practices. (There is one against mutating some of the more fundamental function attributes like `__code__` though.)

Reply:
Ok, that makes a lot of sense. I’ll go read that PEP. This is right up there with the other esoteric stuff Matt P was talking about in the meeting. It is not immediately clear to me what the dangers are of sharing a function object across threads and having threads update attributes concurrent with calls to the function etc. 

FWIW, we have no moral prejudice on hacking either, but we don’t like that situation. 

Thanks!

Reply:
Assuming that by "thread" you mean "subinterpreter": Before a function object is shared between subinterpreters, it should be made transitively immutable and immortal (that's your "immutable" category), otherwise its `__dict__` would be shared mutable state, and that's bad in your model as well as in ours. (In ours, it would allow an object created in one subinterpreter to be deleted in another, and then the allocators get confused because of the way the separate heaps are implemented.)

Reply:
d375a5c6ddb8fe793f16409bacb396e376b2c8e5